core/src/main/java/org/apache/calcite/rel/rules/SemiJoinProjectTransposeRule.java

```diff
@@ -63,42 +63,35 @@
    */
   private SemiJoinProjectTransposeRule(RelBuilderFactory relBuilderFactory) {
     super(
-        operand(SemiJoin.class,
+        operandJ(LogicalJoin.class, null, Join::isSemiJoin,
```

Comment:
I have add the condition `Join::isSemiJoin`, they are thought to be equivalent.

Reply:
Will it still match a SemiJoin.class node?

Reply:
Correct me if I am wrong but people who will still use SemiJoin class or subclasses may not realise that the existing rules have changed and do not match anymore.

Reply:
You are right, but the rule's implementation all match `LogicalJoin` instead of `SemiJoin`, we may need 2 implementations if we want to match `SemiJoin` class. But how can i fix this ? I thought of a way, rename a new Rule class which keep the old logic and make this Rule#Instance reference the old one, but do we really need this ? how about put them in the breaking changes and notes that this rule's match logic has changed.

Reply:
I mostly agree but since `SemiJoin` has a public constructor and factory we cannot be really sure how people are using it. I'm not saying not change it, I'm just a bit defensive and think that a short notice does not harm anybody.

Reply:
Thanks, @zabetak, i would add some notes in the breaking changes for the rule that matches `SemiJoin` class specifically.